File: conformance/tests/post-quantum-wallet/post-quantum-wallet.runar.zig
const runar = @import("runar");

pub const PostQuantumWallet = struct {
    pub const Contract = runar.SmartContract;

    ecdsaPubKeyHash: runar.Addr,
    wotsPubKeyHash: runar.ByteString,

    pub fn init(ecdsaPubKeyHash: runar.Addr, wotsPubKeyHash: runar.ByteString) PostQuantumWallet {
        return .{ .ecdsaPubKeyHash = ecdsaPubKeyHash, .wotsPubKeyHash = wotsPubKeyHash };
    }

    pub fn spend(self: *const PostQuantumWallet, wotsSig: runar.ByteString, wotsPubKey: runar.ByteString, sig: runar.Sig, pubKey: runar.PubKey) void {
        runar.assert(runar.hash160(pubKey) == self.ecdsaPubKeyHash);
        runar.assert(runar.checkSig(sig, pubKey));
        runar.assert(runar.hash160(wotsPubKey) == self.wotsPubKeyHash);
        runar.assert(runar.verifyWOTS(sig, wotsSig, wotsPubKey));
    }
};

File: conformance/tests/post-quantum-wallet/post-quantum-wallet.runar.go
//go:build ignore

package contract

import "runar"

type PostQuantumWallet struct {
	runar.SmartContract
	EcdsaPubKeyHash runar.Addr       `runar:"readonly"`
	WotsPubKeyHash  runar.ByteString `runar:"readonly"`
}

func (c *PostQuantumWallet) Spend(wotsSig runar.ByteString, wotsPubKey runar.ByteString, sig runar.Sig, pubKey runar.PubKey) {
	// Step 1: Verify ECDSA
	runar.Assert(runar.Hash160(pubKey) == c.EcdsaPubKeyHash)
	runar.Assert(runar.CheckSig(sig, pubKey))

	// Step 2: Verify WOTS+
	runar.Assert(runar.Hash160(wotsPubKey) == c.WotsPubKeyHash)
	runar.Assert(runar.VerifyWOTS(sig, wotsSig, wotsPubKey))
}
